listenbrainz/spotify_updater/spotify_read_listens.py

```diff
@@ -0,0 +1,246 @@
+        'tracknumber': track['track_number'],
+        'spotify_artist_ids': [a['external_urls']['spotify'] for a in artists],
+        'artist_names': [a['name'] for a in artists],
+        'listening_from': 'spotify',
+        'disc_number': track['disc_number'],
+        'duration_ms': track['duration_ms'],
+        'spotify_album_id': track['album']['external_urls']['spotify'],
+        # Named 'release_*' because 'release_name' is an official name in the docs
+        'release_artist_name': album_artist_name,
+        'release_artist_names': [a['name'] for a in album_artists],
+        # Named 'album_*' because Spotify calls it album and this is spotify-specific
+        'spotify_album_artist_ids': [a['external_urls']['spotify'] for a in album_artists],
+    }
+    isrc = track.get('external_ids', {}).get('isrc')
+    spotify_url = track.get('external_urls', {}).get('spotify')
+    if isrc:
+        additional['isrc'] = isrc
+    if spotify_url:
+        additional['spotify_id'] = spotify_url
+
+    track_meta = {
+        'artist_name': artist_name,
+        'track_name': track['name'],
+        'release_name': album['name'],
+        'additional_info': additional,
+    }
+
+    return {
+        'listened_at': listened_at,
+        'track_metadata': track_meta,
+    }
+
+
+def get_user_recently_played(user):
+    """ Get recently played songs from Spotify for specified user.
+    This uses the 'me/player/recently-played' endpoint, which only allows us to get the last 50 plays
+    for one user.
+
+    Args:
+        user (spotify.Spotify): the user whose plays are to be imported.
+
+    Returns:
+        the response from the spotify API consisting of the list of recently played songs.
+
+    Raises:
+        spotify.SpotifyAPIError: if we encounter errors from the Spotify API.
+        spotify.SpotifyListenBrainzError: if we encounter a rate limit, even after retrying.
+    """
+    retries = 10
+    delay = 1
+    tried_to_refresh_token = False
+
+    while retries > 0:
+        try:
+            recently_played = user.get_spotipy_client()._get("me/player/recently-played", limit=50)
+            break
+        except SpotifyException as e:
+            retries -= 1
+            if e.http_status == 429:
+                # Rate Limit Problems -- the client handles these, but it can still give up
+                # after a certain number of retries, so we look at the header and try the
+                # request again, if the error is raised
+                time_to_sleep = e.headers.get('Retry-After', delay)
+                current_app.logger.warn('Encountered a rate limit, sleeping %d seconds and trying again...', time_to_sleep)
+                time.sleep(time_to_sleep)
+                delay += 1
+                if retries == 0:
+                    raise spotify.SpotifyListenBrainzError('Encountered a rate limit.')
+            elif e.http_status in (400, 403, 404) or (e.http_status >= 500 and e.http_status < 600):
```

Comment:
The 400 series basically means "caller's fault" and 500 is "server's fault". Are we sure we want to retry 400 errors?? Retying 403 and 404 doesn't really make sense to me...